deployment/legacy-migration/maintenance-worker/migration/neo4j/contributions/contributions.cql

```diff
@@ -125,7 +125,6 @@
 [ ]   wasSeeded: { type: Boolean }
     }
 */
-
 CALL apoc.load.json("file:${IMPORT_CHUNK_PATH_CQL_FILE}") YIELD value as post
 MERGE (p:Post {id: post._id["$oid"]})
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-30T12:18:49Z" title="Friday, August 30th 2019, 2:18:49 pm +02:00">Aug 30, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -148,6 +148,10 @@ MATCH (c:Category {id: categoryId})
 MERGE (p)-[:CATEGORIZED]->(c)
 WITH p, post.tags AS tags
 UNWIND tags AS tag
-MERGE (t:Tag {id: apoc.text.clean(tag), disabled: false, deleted: false})
+WITH apoc.text.replace(tag, '[^\\p{L}0-9]', '') as tagNoSpacesAllowed
+CALL apoc.when(tagNoSpacesAllowed =~ '^((\\p{L}+[\\p{L}0-9]*)|([0-9]+\\p{L}+[\\p{L}0-9]*))$', 'RETURN tagNoSpacesAllowed', '', {tagNoSpacesAllowed: tagNoSpacesAllowed}) 
```


Are you trying to remove hashtags that don't match the regex?

I believe that the empty string `''` is not null, so we import the hashtag in any case.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T12:27:46Z" title="Friday, August 30th 2019, 2:27:46 pm +02:00">Aug 30, 2019</time>_

---


when I was doing the import without this line of code it was failing on line `155` saying that it could not create a `Tag` with `null` as id, so it was coming from somewhere

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-30T12:32:22Z" title="Friday, August 30th 2019, 2:32:22 pm +02:00">Aug 30, 2019</time>_

---


The hashtag should be imported only if:
- after cleaning the hashtag from all not allowed chars:
```
apoc.text.replace(tag, '[^\\p{L}0-9]', ‚‘)
```
- it matches the pattern:
```
'^((\\p{L}+[\\p{L}0-9]*)|([0-9]+\\p{L}+[\\p{L}0-9]*))$'
```

In this new formulation it seems to me that the in front cleaning is not necessary anymore …

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-30T12:34:58Z" title="Friday, August 30th 2019, 2:34:58 pm +02:00">Aug 30, 2019</time>_

---


Ah @mattwr18 replied.
He can answer your question better and tryout, because he has tried out and I can not import my self.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T12:38:19Z" title="Friday, August 30th 2019, 2:38:19 pm +02:00">Aug 30, 2019</time>_

---


that might actually be true @Tirokk... certainly I can test it out, if we don't mind waiting longer for it to get merged in

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-30T12:41:07Z" title="Friday, August 30th 2019, 2:41:07 pm +02:00">Aug 30, 2019</time>_

---


If this solution saves us complexity in the frontend and backend, please remove the obsolete source code! :relieved:

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-30T12:42:05Z" title="Friday, August 30th 2019, 2:42:05 pm +02:00">Aug 30, 2019</time>_

---


Would be great if you could care for it @mattwr18 !!! 😘
I am not in a big hurry. Would be just nice to have it merged before the next production gets deployed. @roschaefer

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T12:49:10Z" title="Friday, August 30th 2019, 2:49:10 pm +02:00">Aug 30, 2019</time>_

---


sure! I'll take care of it today

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-30T13:14:10Z" title="Friday, August 30th 2019, 3:14:10 pm +02:00">Aug 30, 2019</time>_

---


I debugged this with the following:
```
...
WITH p, post, post.tags AS tags
UNWIND tags AS tag
WITH post, apoc.text.replace(tag, '[^a-z^A-Z^0-9]', '') as tagNoSpacesAllowed
CALL apoc.when(tagNoSpacesAllowed =~ '^(([a-zA-Z]+[a-zA-Z0-9]*)|([0-9]+[a-zA-Z]+[a-zA-Z0-9]*))$', 'RETURN tagNoSpacesAllowed', '', {tagNoSpacesAllowed: tagNoSpacesAllowed})
YIELD value as validated
WITH validated, post
WHERE validated.tagNoSpacesAllowed IS NULL
CALL apoc.log.info('Post %s', post)
...
```
then the posts with tags that would become NULL are logged to `/var/log/neo4j/debug.log` (on my machine).

I used MongoDB compass to check out those contributions with tags converted to NULL:

![2019-08-30-151035_1920x1080_scrot](https://user-images.githubusercontent.com/2110676/64023488-b4985500-cb38-11e9-90d9-475932ae9b9e.png)
![2019-08-30-151144_1920x1080_scrot](https://user-images.githubusercontent.com/2110676/64023489-b4985500-cb38-11e9-8be1-fe14004d8d6c.png)
![2019-08-30-151236_1920x1080_scrot](https://user-images.githubusercontent.com/2110676/64023490-b4985500-cb38-11e9-993c-6139032bcf9a.png)

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T13:20:02Z" title="Friday, August 30th 2019, 3:20:02 pm +02:00">Aug 30, 2019</time>_

---


ok, I don't see why they would be converted to null @roschaefer... is it supposed to be obvious?
I just retried the import without the `IS NOT NULL` and it failed as yesterday.

well, I guess the ones that return null would be the ones that after stripped contain only numbers?
do we care about this? I'm personally ok to remove those three tags, what about others?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-30T13:41:43Z" title="Friday, August 30th 2019, 3:41:43 pm +02:00">Aug 30, 2019</time>_

---


Yes, it's because of the numbers. I'm OK to remove these tags, it's just that you could probably use a `COALESCE(somelist)` here.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T14:39:25Z" title="Friday, August 30th 2019, 4:39:25 pm +02:00">Aug 30, 2019</time>_

---


but where is `somelist` coming from? these `tags` only become null as they are passed one-by-one through the regex and then the tags are one-by-one created with their id. we need to skip the `NULL` ones so as not to attempt to create them.